0005-snarkyjs-ecdsa-sha.md

```diff
@@ -0,0 +1,179 @@
+
+## Detailed design
+
+### SHA3/Keccak
+
+The Keccak and SHA3 gadget has been implemented by the crypto team, enabling us to leverage the already existing bindings layer to and from OCaml. This design allows us to simply integrate the new gadgets into SnarkyJS by simply exposing them in `ocaml/lib/snarky_js_bindings_lib.ml`.
+
+For Keccak/SHA3, the implementation exposes two ready-to-go functions.
+
+```ocaml
+(** Gagdet for NIST SHA-3 function for output lengths 224/256/384/512 *)
+val nist_sha3 :
+    (module Snarky_backendless.Snark_intf.Run with type field = 'f)
+  -> int
+  -> 'f Snarky_backendless.Cvar.t list
+  -> 'f Snarky_backendless.Cvar.t array
+
+(** Gadget for Keccak hash function for the parameters used in Ethereum *)
+val eth_keccak :
+    (module Snarky_backendless.Snark_intf.Run with type field = 'f)
+  -> 'f Snarky_backendless.Cvar.t list
+  -> 'f Snarky_backendless.Cvar.t array
+
+```
+
+These two functions will be imported into the bindings layer and exposed via a new sub-module as part of the already existing `Snarky` module.
+
+```ocaml
+
+  module Sha = struct
+    let create message nist length =
+      let message_array = Array.to_list message in
+      if Js.to_bool nist then
+        Kimchi_gadgets.Keccak.nist_sha3 (module Impl) length message_array
+      else Kimchi_gadgets.Keccak.eth_keccak (module Impl) message_array
+  end
+
+```
+
+In order to reduce the bindings surface, Keccak and SHA3 will be exposed via a `create` function, which behaves like a factory pattern.
+By calling this function inside SnarkyJS, we can define and expose all possible variants of SHA3(224/256/384/512) and Keccak without the need to have an individual function in the bindings layer for each variant.
+
+In order to differentiate Poseidon, which works over native Field elements, and SHA3 and Keccak which works over byte-sized Field elements, it will be beneficial to introduce a new type `UInt8` (a Field element that is exactly a byte) to draw a clan line between both hash functions.
+
+In SnarkyJS, these new primitives will be declared as followed:
+
+```ts
+function buildSha(length: 224 | 256 | 385 | 512, nist: boolean) {
+  return {
+    hash(message: UInt8[]) {
+      let payload = [0, ...message.map((f) => f.value)];
+      return Snarky.sha.create(payload, nist, length).map(Field);
+    },
+  };
+}
+
+const Sha3_224 = buildSha(224, true);
+const Sha3_256 = buildSha(256, true);
+const Sha3_385 = buildSha(385, true);
+const Sha3_512 = buildSha(512, true);
+const Keccak = buildSha(256, false);
+```
+
+Another alternative to the factory pattern above could be to supply the developer with a single function that takes a range of parameters, so that the developer can choose their flavour of SHA3/Keccak on their own.
+
+```ts
+function SHA3(
+  length: 224 | 256 | 385 | 512,
+  nist?: boolean = true
```

Comment:
Yeah, perhaps you should state quite clearly in the docs or somewhere else that for now the `nist_version = false` only works if the provided length is 256. Meaning, even if a user wants `Keccak.ethereum` for output length 512, what it will obtain is 256 bit output. Again, let me know if product wants to extend this functionality because it should be straightforward to add.

Reply:
In this case, assuming we don't need the other NIST lengths, we should definitely change the parameters of the function so that only the valid inputs typecheck.

Something like:

```typescript
function SHA3(length: 224 | 256 | 385 | 512 | { nist: 256 }, /*...*)
```